src/utils/publishCustomFunctions.ts

```diff
@@ -0,0 +1,177 @@
+  script.runInNewContext(ctx);
+  const customFunctions = ctx.fn as NamedObject;
+
+  const functionsJsonFile = path.join(workingDir, 'functions.json');
+  const metaData = fs.readJsonSync(functionsJsonFile);
+
+  const groomedMetaData = await Object.keys(customFunctions).reduce(
+    async (promise: Promise<MetaData>, name: string): Promise<MetaData> => {
+      return promise.then(async groomed => {
+        // eslint-disable-next-line no-param-reassign
+        groomed[name] = metaData[name];
+
+        if (!groomed[name]) {
+          const matches = customFunctions[name]
+            .toString()
+            .match(/await context\((["'])(.*?)\1\)/g);
+
+          const defaultInputVariables = (matches || [])
+            .map(m => `${m.slice(15, -2)}:string`)
+            .join(' ');
+
+          // eslint-disable-next-line no-param-reassign
+          groomed = await resolveMissingFunction(
+            groomed,
+            metaData,
+            name,
+            defaultInputVariables,
+          );
+        }
+
+        return groomed;
+      });
+    },
+    Promise.resolve({} as MetaData),
+  );
+
+  fs.writeFileSync(functionsJsonFile, JSON.stringify(groomedMetaData, null, 2));
+
+  return groomedMetaData;
+};
+
+const publishFunctions = async (
+  targetHost: string,
+  metaData: MetaData,
+  bumpRevision: boolean,
+): Promise<void> => {
+  const ide = new IDE(targetHost);
+  const revision = await storeCustomFunctions(ide, metaData, bumpRevision);
+
+  const buildDir = path.join(os.tmpdir(), identifier);
+  const customJsFile = path.join(buildDir, 'dist', 'custom.js');
+
+  await ide.post(
+    `custom_functions/${revision}`,
+    { multiPartData: [{ name: 'code', file: customJsFile }] },
+    `Uploading "${revision}.js" ...`,
+  );
+
+  const actions = (await ide.get('actions')) as Actions;
+
+  await actions.reduce(
+    async (
+      promise: Promise<string | object | null>,
+      { id, use_new_runtime, description }: Action,
+    ) => {
+      await promise;
+      if (use_new_runtime) {
+        return ide.put(
+          `actions/${id}`,
+          { json: { record: { description } } },
+          `Compiling action "${description}" ...`,
+        );
+      }
+      return Promise.resolve(null);
+    },
+    Promise.resolve(null),
+  );
+};
+
+const cleanMetaData = async (): Promise<void> => {
+  const functionsJsonFile = path.join(workingDir, 'functions.json');
+  const metaData = fs.readJsonSync(functionsJsonFile);
+
+  Object.keys(metaData).forEach(name => {
+    delete metaData[name].replace;
+  });
+
+  fs.writeFileSync(functionsJsonFile, JSON.stringify(metaData, null, 2));
+};
+
+const publishCustomFunctions = (
+  host: string,
+  bumpRevision: boolean,
+  skipBuild: boolean,
+): void => {
+  identifier = acquireCustomFunctionsProject(workingDir);
+
+  const targetHost = host || `https://${identifier}.bettyblocks.com`;
+  console.log(`Publishing to ${targetHost} ...`);
+
+  new Promise((resolve): void => {
```

Comment:
Is this promise needed? What does it add? It is only resolved, never rejected. Is it possible to do:

```javascript
    if(!skipBuild) {
      const building = ora(
        `Building custom functions bundle (this can take a while) ...`,
      ).start();
      const build = spawn('bb functions build', {
        shell: true,
      });
      build.on('close', () => {
        building.succeed();
      });
    }
```
?

Reply:
Yes, the promise is needed because this code fragment should be completed before continuing on grooming metadata and so forth.